lightning-persister/src/lib.rs

```diff
@@ -0,0 +1,371 @@
+use lightning::chain::transaction::OutPoint;
+use lightning::util::ser::{Writeable, Readable};
+use std::fs;
+use std::io::Error;
+use std::path::Path;
+
+#[cfg(test)]
+use {
+	bitcoin::{BlockHash, Txid},
+	bitcoin::hashes::hex::FromHex,
+	std::collections::HashMap,
+	std::io::Cursor
+};
+
+#[cfg(not(target_os = "windows"))]
+use std::os::unix::io::AsRawFd;
+
+/// FilesystemPersister persists channel data on disk, where each channel's
+/// data is stored in a file named after its funding outpoint.
+///
+/// Warning: this module does the best it can with calls to persist data, but it
+/// can only guarantee that the data is passed to the drive. It is up to the
+/// drive manufacturers to do the actual persistence properly, which they often
+/// don't (especially on consumer-grade hardware). Therefore, it is up to the
+/// user to validate their entire storage stack, to ensure the writes are
+/// persistent.
+/// Corollary: especially when dealing with larger amounts of money, it is best
+/// practice to have multiple channel data backups and not rely only on one
+/// FilesystemPersister.
+pub struct FilesystemPersister {
+	path_to_channel_data: String,
+}
+
+trait DiskWriteable {
+	fn write(&self, writer: &mut fs::File) -> Result<(), Error>;
+}
+
+impl<ChanSigner: ChannelKeys + Writeable> DiskWriteable for ChannelMonitor<ChanSigner> {
+	fn write(&self, writer: &mut fs::File) -> Result<(), Error> {
+		self.write_for_disk(writer)
+	}
+}
+
+impl FilesystemPersister {
+	/// Initialize a new FilesystemPersister and set the path to the individual channels'
+	/// files.
+	pub fn new(path_to_channel_data: String) -> Self {
+		return Self {
+			path_to_channel_data,
+		}
+	}
+
+	fn get_full_filepath(&self, funding_txo: OutPoint) -> String {
+		let path = Path::new(&self.path_to_channel_data);
+		let mut path_buf = path.to_path_buf();
+		path_buf.push(format!("{}_{}", funding_txo.txid.to_hex(), funding_txo.index));
+		path_buf.to_str().unwrap().to_string()
+	}
+
+	// Utility to write a file to disk.
+	fn write_channel_data(&self, funding_txo: OutPoint, monitor: &dyn DiskWriteable) -> std::io::Result<()> {
+		fs::create_dir_all(&self.path_to_channel_data)?;
+		// Do a crazy dance with lots of fsync()s to be overly cautious here...
+		// We never want to end up in a state where we've lost the old data, or end up using the
+		// old data on power loss after we've returned.
+		// The way to atomically write a file on Unix platforms is:
+		// open(tmpname), write(tmpfile), fsync(tmpfile), close(tmpfile), rename(), fsync(dir)
+		let filename = self.get_full_filepath(funding_txo);
+		let tmp_filename = filename.clone() + ".tmp";
+
+		{
+			// Note that going by rust-lang/rust@d602a6b, on MacOS it is only safe to use
+			// rust stdlib 1.36 or higher.
+			let mut f = fs::File::create(&tmp_filename)?;
+			monitor.write(&mut f)?;
+			f.sync_all()?;
+		}
+		fs::rename(&tmp_filename, &filename)?;
+		// Fsync the parent directory on Unix.
```

Comment:
Otherwise you don't have guarantee that the parent directory logical file is also flushed as a physical copy ? And thus you may still have available on-disk the monitor file but without the directory mapping it's hacky to recover it ?

Reply:
Apparently with some filesystems and mount options, but not all, without this fsync the filesystem won't be able to find your file, because the way that it looks for your file is by asking the containing directory whether it contains X file. (I think usually this'd still be fine, but this extra fsync is for mitigating a program crash in the middle of persistence.)

Found more info here: https://lwn.net/Articles/457667/ "When should you fsync"

Reply:
Thanks for the reference :)